### ISSUE TITLE
Any plan on continuing development?

### ISSUE THREAD

**Issue by awesomebytes** [chidea/pyTelegram#1]
Hello, I'm interested in a Python API for telegram, everyone seems to just go with the ugly (and buggy in my tests) CLI-based python API.

I could probably help on the development.

**Comment by chidea** [comment]
Welcome and thank you, Sammy.

As you said, features like that of CLI based telegram is the first goal I'd like to accomplish but still I'm way unknown how telegram protocol works.

I'm reading source code of the official c++ version and only guessing that lower TL-layers could and should be achieved easily before any other layers and features.

By the way, there should be any telecommunication libraries chosen which is best for this project but still, I'm way unknown on this area. Only a thing I know is that Python 3's preincluded socket library is hard to use cuz that's what I've tried in another project.

Please, give me any suggestions or commits. I'll be openly listen and get to dig more.

**Comment by awesomebytes** [comment]
Hey chidea,

I've forked some projects that seem to have the same aim: 
https://github.com/niomaster/python-telegram
https://github.com/griganton/telepy

Then there is this other that seems to try to do some adaptation from the C API:
https://github.com/gokceneraslan/pytelegram
But I think that's not too cool.

So, the thing is, I haven't really understood what is needed to be implemented. Their architecture is... complicated, for giving it a friendly name (or I just haven't seen a plain simple examplanation on how it works). I saw you had some structure started that is maybe the way to go. Also there is this other file: https://github.com/griganton/telepy/blob/master/mtproto.py which also seems to attempt to implement some types.

From my not-so-well-understood idea I think there should be some automated way of converting from the TL-schema:
https://core.telegram.org/schema
(Probably from the JSON version for ease of parsing) to python and even maybe something to do with the methods (are they the same? no, right? https://core.telegram.org/methods).

For example, I don't really understand what does this description mean.
https://core.telegram.org/method/messages.sendMessage

Do we need to have the types that are said in there to use, so when you actually use some sendMessage function, you just translate that as the header with it's defined #code, then the InputPeer will be translated to another #code, and so on with all the parameters?

And finally some networking-like library of the mtp protocol will just send all this in some connection?

Sorry if my explanation sucked! If you would to get in touch in other means (some chat?) I would find it awesome.

**Comment by chidea** [comment]
Wow, you did great research.

I've read #griganton 's code just now and found its cool TL-Schema reading
implementation and it's utilizing socket..!!

Yes, there's very easy way with C API but I agree with you that having pure
python implementation would be more awsome and it will help people
understand telegram better than with C or C++ code.

Yes they are different. Method is API's method and calling them will
just utilize on C so it needs the same C structure described on docs.

What I'm interested in is using pypy to replace C API with easier python
code which is still as much fast as C. But I've found there's bug on
windows with code page which will be fixed on its next version, so I'm
waiting for it.

Those structures I've written actually came from C++ version of code.
There's very low level implments there deserve reading but just hard to.

Maybe we should use gitter.im to chat with.
Here's URL : https://gitter.im/chidea/pyTelegram

**Comment by awesomebytes** [comment]
@griganton you should join this conversation and we should unify efforts!

As of implementation details... I would aim for a Python 2.7 implementation (which should work on Python 3 by default right?) and don't bother much with its speed for now. Just get a working implementation first. In my experience Pypy doesn't really execute everything. Using Cython can speed up things without any effort (other than setting a script to convert to C and compile to shared lib).

I'm all in on helping on all this matter.

**Comment by superchromia** [comment]
Hello, everyone. I was surprized that my project interested you ! ))

About Telegram. It uses its own protocol called "MTProto". First step is to implement it. 
It's all about binaries. It is not XML-like XMPP, so we have to learn how to serialize/deserialize data.

And we have no chance to implement TL schema by hardcoding. Resulting TL is huge, so we have to parse it, create functions and objects dynamically.

I've started to implement steps from here: https://core.telegram.org/mtproto/samples-auth_key.
Today I'm on step 4. =)

 /home/anton/Projects/telepy/testing.py

Sent:
78 97 46 60 DC 9B 59 A0
69 90 D3 CD 24 F8 87 2C
3C DC E3 82

Received:
63 24 16 05 DC 9B 59 A0
69 90 D3 CD 24 F8 87 2C
3C DC E3 82 D3 9A 2E 7A
F4 9A 46 AE 64 06 E8 1F
B5 4F CE AD 08 1C 20 24
CD 06 29 C2 95 00 00 00
15 C4 B5 1C 01 00 00 00
21 6B E8 6C 02 2B B4 C3

PQ = 2026660295307018901
p = 1327666733, q = 1526482697

Process finished with exit code 0.

**Comment by superchromia** [comment]
No, Python 3.4 and Python 2.7 are not compatible at all... So I prefer to use Python3.
Just to get a working implementation first.

**Comment by awesomebytes** [comment]
Uhm, I can't work on Python 3.4. I don't have any system that uses it. Are you familiar with ways of making code compatible for both? I can do research on it. (I'm all in on making it compatible for both versions! but having an initial working version is probably the most important thing, as it can attract users that may help on building a robust API).

**Comment by Nekmo** [comment]
I am also interested in this project :)
